shap/_explanation.py

```diff
@@ -170,7 +172,7 @@ def __init__(
         )
 
     @property
-    def shape(self) -> tuple[Union[int, None], ...]:
+    def shape(self) -> tuple[int | None, ...]:
```

Comment:
this won't work well on python3.9. I think we could drop support for python3.9 altogether though.

Reply:
On a related note, I found two neat ruff linting rules which we can enable for consistent use of type hints:
- [FA](https://docs.astral.sh/ruff/rules/#flake8-future-annotations-fa): future annotations
- [TCH](https://docs.astral.sh/ruff/rules/#flake8-type-checking-tch): type checking

I pushed them to this branch, which required a few small modifications to get them passing.